### PR TITLE
Ensembles

### LINKED ISSUE #3: Generate ensemble pipelines with an arbitrary number of ensembled subpipelines

### PR DESCRIPTION
**Pull request by orionw**
This PR is after the Validation PR
- closes #3 

Features:
- Generates ensembles with a arbitrary number of sub-pipelines, with options for random and same-model predictors, random and same-model preprocessors, as well as specific models and model order.
- Test cases for generating ensembles and error checking
- Test cases for executing ensembles and simulating the whole process

We still need to decide how we want to use these: should I add a command line option?  Should we just do it manually?

### CODE REVIEW

File: experimenter/experimenter.py [hunk 1]
```diff
@@ -64,14 +65,21 @@ def __init__(self, datasets_dir: str, volumes_dir: str, input_problem_directory=
...
                                  len(self.generated_pipelines["regression"])
 
             print("There are {} pipelines".format(self.num_pipelines))
 
             if location is None:
+                self.mongo_database = PipelineDB()
                 print('Exporting pipelines to mongodb...')
                 self.output_pipelines_to_mongodb()
+            elif location == "test":
+                pass
```
Comment (bjschoenfeld) [inline experimenter/experimenter.py]: This does not seem like good practice to have test related code in real code. You can set location to a specific place and then clean it up after the test is over.

Comment (orionw) [reply experimenter/experimenter.py]: The problem is that one of the test cases in test_generate_pipelines was creating a pipeline and then submitting it to mongodb.  I wanted to have an option to generate pipelines but not store them anywhere.  Perhaps a better name would be "NoStore" or something like that.  We could pass it a location for the test, but I thought that case deserved it's own option.  I'm fine either way though.

Comment (bjschoenfeld) [reply experimenter/experimenter.py]: > The problem is that one of the test cases in test_generate_pipelines was creating a pipeline and then submitting it to mongodb.

Can we write the pipeline to a dummy location and then delete it later?

Comment (orionw) [reply experimenter/experimenter.py]: Sure that works!  I just thought maybe we wanted an option to generate but not store right away.  I can't think of many uses offhand, so we can definitely go that route

Comment (bjschoenfeld) [reply experimenter/experimenter.py]: Ideally (I think) we would not be writing to any location in the constructor. Rather, we would have methods to write to database or to write to file.

Comment (orionw) [reply experimenter/experimenter.py]: Yes, I agree.  Do you want me to refactor it for this PR?

Comment (bjschoenfeld) [reply experimenter/experimenter.py]: Let's not do the major refactoring for this PR. Let's just write the files to /dev/null for tests

File: experimenter/constants.py [hunk 2]
```diff
@@ -96,6 +94,50 @@
...
+    "uu3_world_development_indicators",
+    "LL0_344_mv",
+    "LL0_574_house_16h",
+    "LL0_296_ailerons",
+    "LL0_216_elevators",
+    "LL0_201_pol",
+    "uu2_gp_hyperparameter_estimation_v2",  # has extra data
+    "uu2_gp_hyperparameter_estimation",  # has extra data
+    "57_hypothyroid",  # always NAN's out
+
```
Comment (bjschoenfeld) [inline experimenter/constants.py]: do we need all these blank lines?

Comment (orionw) [reply experimenter/constants.py]: Nope. Deleting them

File: docker-compose.yml [hunk 3]
```diff
@@ -12,6 +12,7 @@ services:
       - ${VOLUMES}:/volumes
       - ${DATASETS}:/datasets
       - ${EXPERIMENTER}:/d3m-experimenter
+      - ${D3M}:/d3m
```
Comment (bjschoenfeld) [inline docker-compose.yml]: why is this needed? it is installed in the base image

Comment (orionw) [reply docker-compose.yml]: Yeah I think I was tired of adding it into all my branch work and then just decided to leave it.  I'll take it out

File: experimenter/experimenter.py (hunk 1 above)
Comment (orionw) [reply experimenter/experimenter.py]: Do I need to change anything in this PR for this?  I'm trying to get this ready so we can merge it in

Comment (bjschoenfeld) [reply experimenter/experimenter.py]: Yes, let's remove `location == "test"`. Us the path "/dev/null" in the test code when you don't want it to go anywhere.

File: tests/test_pipeline_generator.py [hunk 4]
```diff
@@ -20,7 +20,7 @@ def setUp(self):
         self.experimenter_driver = Experimenter(
             self.datasets_dir, volumes_dir,
             self.seed_problem_directory, models, preprocessors,
-            generate_problems=True
+            generate_problems=True, location="test"
```
Comment (bjschoenfeld) [inline tests/test_pipeline_generator.py]: Let's remove this, as above.

Comment (orionw) [reply tests/test_pipeline_generator.py]: It is removed!
